---
title: "Using the DMR Scan Package"
author: "Christian M Page"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{DMRScan Vignette}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, echo = FALSE}
library(DMRScan)
```
# Abstract
###Motivation
DNA methylation plays an important role in human health and disease, and methods for the identification of differently methylated regions are of    increasing interest. There is currently a lack of statistical methods which properly address multiple testing, i.e. genome-wide significance for    differentially methylated regions.
###Methods
We introduce a scan statistic (DMRScan), which overcomes these limitations. We benchmark DMRScan against two well established methods (Bumphunter,  DMRcate), using a simulation study based on real methylation data.
###Results
Our method had a higher power than alternative methods across different simulation scenarios, particularly for small effect sizes. DMRScan exhibits greater flexibility in statistical modeling and can be used with more complex designs than current methods.
###Conclusion
DMRScan is the first dynamic approach which properly addresses the multiple-testing challenges for the identification of differently methylated     regions. DMRScan outperformed alternative methods in terms of power, while keeping the false discovery rate controlled.

## Data Set
Included in this package is an example data set. This data set is an extraction of chromosome 22 from a RRBS from the FinHinT study. Please  


# Examples of use of DMRScan

```{r}
data(DMRScan.methylationData) ## Load methylation data from chromosome 22
data(DMRScan.phenotypes) ## Load phenotype (end-point for methylation data)

## Test for an association between phenotype and Methylation
test.statistics <- apply(DMRScan.methylationData,1,function(x,y)summary(glm(y ~ x, family = binomial(link = "logit")))$coefficients[2,3], y = DMRScan.phenotypes)

pos <- data.frame(matrix(as.integer(unlist(strsplit(names(test.statistics),split="chr|[.]"))), ncol = 3, byrow = TRUE))[,-1] ## Set chromosomal position to each test-statistic

## Set clustering features 
min.cpg <- 3  ## Minimum number of CpGs in a tested cluster
max.gap <- 750  ## Maxium distance (in base-pairs) within a cluster before it is broken up into two seperate cluster 

window.sizes <- 3:7 ## Number of CpGs in the sliding windows (can be either a single number or a sequence)
n.CpG        <- nrow(DMRScan.methylationData) ## Number of CpGs to be tested

## Estimate the window threshold, based on the number of CpGs and window sizes
window.thresholds <- estimate_t_grid(k_grid = window.sizes, L = n.CpG, method = "zhang", mcmc = 10000)

## Identify all clusters, and generate a list for each cluster
regions <- make.cpg.regions(test_statistic = test.statistics, chr = pos[,1], pos = pos[,2], max.gap = max.gap, min.cpg = min.cpg)
## Run the sliding window
dmrscan.results   <- dmr_scan(obs = regions, k_grid = window.sizes, t_grid = window.thresholds)
## Print the result
print(dmrscan.results)

```
This will give the following result, with four genome wide significant regions. Two DMRs containing 3 CpGs and two containing 5 CpGs. 

| Genomic Coord. | Start |  End  |Length | Bump |   p.val.emp |           Pos          |
|----------------|-------|-------|-------|-----:|-------------|-------------------------
| chr22.51002050 | 51043 | 51045 |     3 | 2539 | 1.056161e-05| chr22:51002050-51002059|
| chr22.23801586 | 10807 | 10809 |     3 | 573  | 2.112323e-05| chr22:23801586-23801597|
| chr22.43821030 | 36791 | 36795 |     5 | 1925 | 2.217939e-04| chr22:43821030-43821098|
| chr22.46762751 | 41946 | 41950 |     5 | 2158 | 2.693212e-04| chr22:46762751-46762782|


When using a more stringent cut-off, generated by using the "_siegmund_" option in `dmr_scan()`, we get no significant regions on the same data set. Exemplified by the syntax below:

```{r}
window.thresholds <- estimate_t_grid(k_grid = window.sizes, L = n.CpG, method = "siegmund")

dmrscan.results   <- dmr_scan(obs = regions, k_grid = window.sizes, t_grid = window.thresholds)

## Print the result
print(dmrscan.results)
```
Which will return an object of type 'NA', since no genome wide significant regions were found.

End of vignette
